sdk_refactor_design_proposal.md

```diff
@@ -0,0 +1,392 @@
+      - [Flow](#flow)
+      
+    + [Limitations in current SDK Flow](#limitations-in-current-sdk-flow)
+        * [Client constructor](#client-constructor)
+        * [Proposed flow](#proposed-flow)
+        * [API and commands](#api-and-commands)
+        * [Proposed flow](#proposed-flow-1)
+        
+    + [Full proposed SDK flow](#full-proposed-sdk-flow)
+    
+    + [Backwards compatibility](#backwards-compatibility)
+    
+    + [Delivery plan](#delivery-plan)
+    
+        
+
+### Resources
+
+| Resource                  | Link                                      |
+| ------------------------- | ----------------------------------------- |
+| Zen of Python (the Bible) | https://www.python.org/dev/peps/pep-0020/ |
+
+
+
+### Motivation
+
+While GAing the CLI, it was realized that the SDK has the following limitations, making the code not easily readiable and hard to make additions:
+
+1. Classes handle multiple responsibilities, impeding on the ability to cover current and future usecases without expensive refactoring
+2. Implicit instead of explicit, forcing the end-user to understand the inner workings of our Client and the context they are running in, instead of allowing them to explicitly accomplish their desired flow. See [Zen of Python](https://www.python.org/dev/peps/pep-0020/) for more details.
+3. Areas of volatility are not encapsulated, becomes problematic when the requirements change or additions need to be made.
+
+In this document, I will go into more detail for each of these points as well as supply a proposal for how to address them. 
+
+It is important to note that this refactor will help us in our effort to decouple the CLI logic from the SDK. That way, when we are ready and decide to make that separation, we can do so more easily.
+
+
+
+### Current SDK implementation
+
+#### Flow
+
+The below image details the calls from when a user initalizes the *current* Client to when they receive a response.
+
+![CurrentSDKFlow](design/assets/CurrentSDKFlow.png)
+
+At a high level,
+
+1. The user initializes the *Client* with details needed to connect to the *Conjur server*
+
+    `client = conjur.Client(url="https://someserver", account="cucumber", password="123", ...)`.
+
+2. The *Client* constructor is called, ensuring all configuration and credential information needed to make a request have been passed in. If not, it fetches the missing information in a discovery-like manner. For example, for `login`, if a Keystore is not available, it will attempt to fetch credentials from the netrc. At this point the *API* is also initalized.
+
+3. *Client* is now initalized and the user can now make a request.
+
+   `list_values = client.list()`
+
+4. Once a request is made, the *Client* makes a call to *API* which handles the building of the request, merging parameters together, understanding which endpoint to use, etc.
+
+5. The *API* calls the *HTTPWrapper* which escapes the parameters and invokes the endpoint to the *Conjur Server*.
```

Comment:
What methods are we using to sanitize those parameters? Can you give an example for before and after?

Reply:
Not sure yet. Currently though we escape. See [here](https://github.com/cyberark/conjur-api-python3/blob/main/conjur/wrapper/http_wrapper.py#L46)

Any suggestions on what to introduce?